### PR TITLE
#941 added subphase support to loadComm

### LINKED ISSUE #941: Record communication stats per-subphase
**What Needs to be Done?**

- [ ] Extend the instrumentation of objects' communication loads to record them at subphase granularity
- [ ] Update the `CommOverhead` load model to use this

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by bradybray**
fixes #941 implementing subphases in loadComm

### CODE REVIEW

**Comment by PhilMiller** [comment]
So, to carry this the rest of the way through, we'll need a per-subphase structure alongside `node_comm_` in `node_stats.h`, and appropriate recording into it in `NodeStats::addNodeStats`. That means `ElementStats` will need to record it in `ElementStats::recv` to provide the data later

**Comment by bradybray** [comment]
@PhilMiller 
> So, to carry this the rest of the way through, we'll need a per-subphase structure alongside `node_comm_` in `node_stats.h`, and appropriate recording into it in `NodeStats::addNodeStats`. That means `ElementStats` will need to record it in `ElementStats::recv*` to provide the data later

https://github.com/DARMA-tasking/vt/blob/1a4c8bc2eafcf3a3455b88a680772bd2bfe4ff6d/src/vt/vrt/collection/balance/node_stats.h#L257-L267

So I see there is a node_subphase_data_ structure on line 257. Is this not a sufficient structure for this implementation? 

Ah, as I wrote this I saw something on line 267. Would 
`std::unordered_map<SubphaseType, CommMapType> node_comm_s;` be sufficient for the structure? 
I'll start peeking into the other files to provide the additional implementation. I'm sure I'll have some more comments to follow ;)

**Comment by lifflander** [comment]
> Ah, as I wrote this I saw something on line 267. Would
> `std::unordered_map<SubphaseType, CommMapType> node_comm_s;` be sufficient for the structure?
> I'll start peeking into the other files to provide the additional implementation. I'm sure I'll have some more comments to follow ;)

I think you will  to refactor to be `std::unordered_map<PhaseType, std::unordered_map<SubphaseType, CommMapType>> node_comm_; `

File: src/vt/vrt/collection/balance/elm_stats.cc [hunk 1]
```diff
@@ -182,25 +189,38 @@ TimeType ElementStats::getLoad(PhaseType phase, SubphaseType subphase) const {
 }
 
 CommMapType const&
-ElementStats::getComm(PhaseType const& phase) {
+ElementStats::getComm(PhaseType const& phase, SubphaseType const& subphase) {
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/elm_stats.cc:192]: Let's get rid of this unused `subphase` argument

File: src/vt/vrt/collection/balance/elm_stats.cc [hunk 2]
```diff
@@ -182,25 +189,38 @@ TimeType ElementStats::getLoad(PhaseType phase, SubphaseType subphase) const {
...
   comm_.resize(phase + 1);
   auto const& phase_comm = comm_[phase];
 
   vt_debug_print(
     lb, node,
-    "ElementStats: getComm: comm size={}, phase={}\n",
-    phase_comm.size(), phase
+    "ElementStats: getComm: comm size={}, phase={}, subphase={}\n",
+    phase_comm.size(), phase, subphase
```
Comment (bradybray) [inline src/vt/vrt/collection/balance/elm_stats.cc:199]: @PhilMiller Its used here in the debug print. Should I remove it from this section as well?

Comment (PhilMiller) [reply src/vt/vrt/collection/balance/elm_stats.cc:199]: Yeah, please remove it entirely. It was added by an earlier commit in this branch, that was trying to address the issue from a different direction than the one we're taking.

**Comment by bradybray** [comment]
@lifflander @PhilMiller `/usr/bin/ld: final link failed: No space left on device` This directory gets full?

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/balance/node_stats.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/balance/node_stats.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  2
- tests/unit/collection/test_lb.extended.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by PhilMiller** [comment]
It looks like the last commit, "Checkpoint before rebase", reverts all of the intended changes in the rest of the PR.

**Comment by bradybray** [comment]
> 
> 
> It looks like the last commit, "Checkpoint before rebase", reverts all of the intended changes in the rest of the PR.

Ok. Do I revert back to the previous commit than push?

**Comment by PhilMiller** [comment]
I think you would want to `git reset` to the previous commit

**Comment by bradybray** [comment]
> 
> 
> I think you would want to `git reset` to the previous commit

Yep that is what I ended up doing. I force pushed the commit.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_lb.extended.cc  2
- tests/unit/collectives/test_mpi_collective.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by bradybray** [comment]
> 
> 
> I think you would want to `git reset` to the previous commit

As for the conflicts; which side of the conflict do I take? I assume I should be taking the side from develop and not my own branch? The conflicts appear to be from the runtime library.

**Comment by PhilMiller** [comment]
There shouldn't be conflicts at all - your changes don't touch those files.

File: src/vt/vrt/collection/balance/node_stats.h [hunk 3]
```diff
@@ -108,13 +108,14 @@ struct NodeStats : runtime::component::Component<NodeStats> {
    * \param[in] phase the current phase
    * \param[in] time the time the object took
    * \param[in] comm the comm graph for the object
-   *
+  eric *
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/node_stats.h:111]: This is a stray insertion, to be cleaned up.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/collectives/test_mpi_collective.cc  2
- src/vt/vrt/collection/manager.impl.h  2
- tests/unit/termination/test_termination_action_common.cc  1
- src/vt/vrt/collection/send/sendable.impl.h  6
- tests/unit/collection/test_lb.extended.cc  2
- tests/unit/collection/test_collection_group.extended.cc  4
- tests/unit/termination/test_termination_action_callable.extended.cc  1
         

Clones removed
==============
+ tests/unit/pipe/test_callback_send.cc  -1
+ examples/collection/jacobi1d_vt.cc  -1
+ tests/unit/pipe/test_callback_func.cc  -1
+ tests/unit/pipe/test_callback_bcast.cc  -1
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
+ src/vt/vrt/collection/balance/node_stats.cc  -2
+ tests/unit/pipe/test_callback_func_ctx.cc  -1
+ examples/collection/jacobi2d_vt.cc  -1
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/termination/test_termination_action_common.cc  1
- tests/unit/collection/test_lb.extended.cc  2
- tests/unit/termination/test_termination_action_callable.extended.cc  1
- src/vt/vrt/collection/manager.impl.h  2
- src/vt/vrt/collection/send/sendable.impl.h  6
- tests/unit/collectives/test_mpi_collective.cc  2
         

Clones removed
==============
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
+ tests/unit/pipe/test_callback_func.cc  -1
+ tests/unit/pipe/test_callback_bcast.cc  -1
+ examples/collection/jacobi1d_vt.cc  -1
+ src/vt/vrt/collection/balance/node_stats.cc  -2
+ tests/unit/pipe/test_callback_send.cc  -1
+ tests/unit/pipe/test_callback_func_ctx.cc  -1
+ examples/collection/jacobi2d_vt.cc  -1
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/send/sendable.impl.h  6
- tests/unit/termination/test_termination_action_common.cc  1
- src/vt/vrt/collection/manager.impl.h  2
- tests/unit/collection/test_lb.extended.cc  2
- tests/unit/collectives/test_mpi_collective.cc  2
- tests/unit/termination/test_termination_action_callable.extended.cc  1
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/collection/test_collection_group.extended.cc  4
         

Clones removed
==============
+ tests/unit/pipe/test_callback_bcast.cc  -1
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
+ examples/collection/jacobi2d_vt.cc  -1
+ tests/unit/pipe/test_callback_func_ctx.cc  -1
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
+ src/vt/vrt/collection/balance/node_stats.cc  -2
+ tests/unit/pipe/test_callback_func.cc  -1
+ tests/unit/pipe/test_callback_send.cc  -1
+ examples/collection/jacobi1d_vt.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_collection_group.extended.cc  4
- tests/unit/collection/test_lb.extended.cc  2
- src/vt/vrt/collection/manager.impl.h  2
- src/vt/vrt/collection/send/sendable.impl.h  6
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/termination/test_termination_action_callable.extended.cc  1
- tests/unit/collectives/test_mpi_collective.cc  2
- tests/unit/termination/test_termination_action_common.cc  1
         

Clones removed
==============
+ tests/unit/pipe/test_callback_send.cc  -1
+ src/vt/vrt/collection/balance/node_stats.cc  -2
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
+ examples/collection/jacobi1d_vt.cc  -1
+ examples/collection/jacobi2d_vt.cc  -1
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
+ tests/unit/pipe/test_callback_func_ctx.cc  -1
+ tests/unit/pipe/test_callback_bcast.cc  -1
+ tests/unit/pipe/test_callback_func.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6221563&bid=20279982)

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**